### PR TITLE
Add new custom channels itest liquidity edge case for safe HTLC failure

### PR DESCRIPTION
**Pull request by GeorgeTsagk**
## Description

This new custom channels liquidity edge case tests that an asset HTLC with more units than the local balance may not get added to the channel.

 Based on 
https://github.com/lightninglabs/taproot-assets/pull/1462
https://github.com/lightningnetwork/lnd/pull/9687

### CODE REVIEW

**Review by guggero — CHANGES_REQUESTED** [review]
Don't thing we're asserting what we really want to assert.

File: itest/litd_custom_channels_test.go [hunk 1]
```diff
@@ -2163,6 +2163,41 @@ func testCustomChannelsLiquidityEdgeCases(ctx context.Context,
...
 	logBalance(t.t, nodes, assetID, "after small manual rfq")
+
+	// Edge case: Fabia creates an invoice which Erin cannot satisfy with
+	// his side of asset liquidity. This tests that Erin will not try to
+	// add an HTLC with more asset units than what his local balance is. To
+	// validate that the channel is still healthy, we follow up with a
+	// smaller invoice payment which is meant to succeed.
+
+	// We now create a hodl invoice on Fabia, for 125k assets.
+	hodlInv = createAssetHodlInvoice(t.t, erin, fabia, 125_000, assetID)
```
[the remark this answers is not in the document]
Comment (guggero) [reply itest/litd_custom_channels_test.go:2174]: It's mostly because the MPP algorithm will try to split it up until it hits the timeout.
So with a bit of creativity we can make this work:

```diff
diff --git a/itest/assets_test.go b/itest/assets_test.go
index 435c5f83..20cbd5cb 100644
--- a/itest/assets_test.go
+++ b/itest/assets_test.go
@@ -935,6 +935,7 @@ func payInvoiceWithSatoshiLastHop(t *testing.T, payer *HarnessNode,
 
 type payConfig struct {
        smallShards       bool
+       noShards          bool
        errSubStr         string
        allowOverpay      bool
        feeLimit          lnwire.MilliSatoshi
@@ -962,6 +963,12 @@ func withSmallShards() payOpt {
        }
 }
 
+func withNoShards() payOpt {
+       return func(c *payConfig) {
+               c.noShards = true
+       }
+}
+
 func withPayErrSubStr(errSubStr string) payOpt {
        return func(c *payConfig) {
                c.errSubStr = errSubStr
@@ -1032,6 +1039,10 @@ func payInvoiceWithAssets(t *testing.T, payer, rfqPeer *HarnessNode,
                sendReq.MaxShardSizeMsat = 80_000_000
        }
 
+       if cfg.noShards {
+               sendReq.MaxParts = 1
+       }
+
        var rfqBytes []byte
        cfg.rfq.WhenSome(func(i rfqmsg.ID) {
                rfqBytes = make([]byte, len(i[:]))
diff --git a/itest/litd_custom_channels_test.go b/itest/litd_custom_channels_test.go
index c15dc59b..e4f7e5d8 100644
--- a/itest/litd_custom_channels_test.go
+++ b/itest/litd_custom_channels_test.go
@@ -2170,28 +2170,19 @@ func testCustomChannelsLiquidityEdgeCases(ctx context.Context,
        // validate that the channel is still healthy, we follow up with a
        // smaller invoice payment which is meant to succeed.
 
-       // We now create a hodl invoice on Fabia, for 125k assets.
-       hodlInv = createAssetHodlInvoice(t.t, erin, fabia, 125_000, assetID)
+       // We now create a large invoice on Fabia, for 125k assets.
+       largeInvoice := createAssetInvoice(t.t, erin, fabia, 125_000, assetID)
 
        // Charlie tries to pay, this is not meant to succeed.
        payInvoiceWithAssets(
-               t.t, charlie, dave, hodlInv.payReq, assetID,
-               withFailure(lnrpc.Payment_IN_FLIGHT, failureNone),
-       )
-
-       payHash = hodlInv.preimage.Hash()
-       _, err = fabia.InvoicesClient.CancelInvoice(
-               ctx, &invoicesrpc.CancelInvoiceMsg{
-                       PaymentHash: payHash[:],
-               },
+               t.t, charlie, dave, largeInvoice.PaymentRequest, assetID,
+               withNoShards(),
+               withFailure(lnrpc.Payment_FAILED, failureNoRoute),
        )
-       require.NoError(t.t, err)
 
        // Now let's create a smaller invoice and pay it, to validate that the
        // channel is still healthy.
-       invoiceResp = createAssetInvoice(
-               t.t, erin, fabia, 100, assetID,
-       )
+       invoiceResp = createAssetInvoice(t.t, erin, fabia, 100, assetID)
 
        payInvoiceWithAssets(
                t.t, charlie, dave, invoiceResp.PaymentRequest, assetID,
```

This can probably also be used further up, where we currently wait for the full 60 second timeout.

[the remark this answers is not in the document]
Comment (guggero) [reply itest/litd_custom_channels_test.go:2174]: Argh... Right. Too much context switching, sorry.

So can we just limit the number of parts then? To cause the pathfinding algorithm to give up more quickly?

File: itest/litd_custom_channels_test.go [hunk 2]
```diff
@@ -3068,7 +3068,7 @@ func testCustomChannelsLiquidityEdgeCasesCore(ctx context.Context,
 	res, err := charlieTap.RfqClient.AddAssetBuyOrder(
 		ctx, &rfqrpc.AddAssetBuyOrderRequest{
 			AssetSpecifier: &assetSpecifier,
-			AssetMaxAmt:    10_000,
+			AssetMaxAmt:    1_000,
```
Comment (guggero) [inline itest/litd_custom_channels_test.go:3071]: nit: comment block above this needs to be updated with the new numbers.

**Review by guggero — APPROVED** [review]
Thanks for the update, LGTM :tada:

**Comment by guggero** [comment]
Rebased and ready for review, @ffranr. This should shave off about 2 minutes of wait time from our custom channel itests.

**Review by ffranr — APPROVED** [review]
I've fixed the lint error and force pushed.

**Comment by guggero** [comment]
> I found an easier way to shave off those extra seconds, simply by lowering the amounts of the quote/invoice leading to a faster failure.

That seems to have introduced a new flake. Going to investigate.
